### ISSUE TITLE
compile error 

### ISSUE THREAD

**Issue by caotangyu** [DentOpt/denmpc#1]
catkin_make output error :  Scheduler.h:17: error: ros/ros.h: No such file or directory

**Comment by DentOpt** [comment]
Hi, on a standard Indigo distribution, it compiles without problems.
The ros/ros.h is not included in the package, but comes with your ROS distribution.
ros/ros.h is normally directly linked to the DenOpt package by the entries:

> catkin_package( CATKIN_DEPENDS roscpp )              # in In the CmakeLists 
> <build_depend>roscpp</build_depend>                       # in In the Package.xml
> <run_depend>roscpp</run_depend>                           # in In the Package.xml

Which ROS version are you using?

**Comment by caotangyu** [comment]
Hi, I am using kinetic version. I add“ include_directories( ${catkin_INCLUDE_DIRS}  ) “to the CmakeLists.txt file.  Than it compiles correctely. May be the CMaLists.txt file under denmpc/src directory is useless,  I delete it.

祝您身体健康，工作顺利

________________________________
发件人: DentOpt <notifications@github.com>
发送时间: Thursday, November 16, 2017 8:15:41 PM
收件人: DentOpt/denmpc
抄送: will1991; Author
主题: Re: [DentOpt/denmpc] compile error (#1)


Hi, on a standard Indigo distribution, it compiles without problems.
The ros/ros.h is not included in the package, but comes with your ROS distribution.
It is should be directly linked in the DenOpt package by the entries:

catkin_package( CATKIN_DEPENDS roscpp ) # in In the CmakeLists
<build_depend>roscpp</build_depend> # in In the Package.xml
<run_depend>roscpp</run_depend> # in In the Package.xml

Which ROS version are you using?

―

**Comment by DentOpt** [comment]
Thank you. 
You are completely right!
The CmakeLists under /src is not used, I deleted it from the Master branch
and also added the ${catkin_INCLUDE_DIRS}.
I do currently not have a kinetic distribution to test it, 
but hope it is resolved now.

**Comment by caotangyu** [comment]
You are welcome.
Recently I am reading your code and paper.
I am confused that the dimension of state is six described in the paper<http://orbilu.uni.lu/bitstream/10993/28640/1/DENTLER_jan_MSC16_corrected.pdf>, but the program set dim_x_ = 9.
Do you have a draft of tutorial ?
Could you tell  me how to run the quadrotors simulation ? I would like test the program under kinetic version.

**Comment by DentOpt** [comment]
Hi Will,
you have the right to be confused :).
- I am trying to find some time to write a proper tutorial with the turtlebot.
The MSC paper has been issued to explaining the structure and implementational details of the DENMPC package with the example application controlling the AR.Drone using
a memory reduced model with dimension dim_x_=6 . 
The published version of the DENMPC package does include a generic model with dim_x_=9
to show the exemplary outline how an agent is implemented in the framework.

So in order to help you properly, the question is what exactly you want to do:

1. Do you want to control a real AR.Drone?
2. Do you want to control a different type of drone?
3. Do you want to control a different type of system (e.g. different kind of robots, predator–prey systems, financial systems, etc.)?
4. You want to simulate the system dynamics of the AR.Drone?
5. You want to simulate the closed-loop behavior of the AR.Drone with MPC?

To give you an easy introduction,
I created a turtlesim scenario.
You can find the simple description under
`https://github.com/DentOpt/denmpc/wiki/DENMPC-Turtlesim-Tutorial`


Best regards,
Jan

**Comment by caotangyu** [comment]
Hi Jan,
Than you for your patience in answering me.
I want use your package to control a real robot that different with AR.Drone . The robot is a race car like this http://www.barc-project.com/.

**Comment by DentOpt** [comment]
Hi, then I would recommend you to take the new tutorial_turtlesim branch, I opened.
There you can basically see, how a similar robot (the turtle) is steered.
If you look at turtle.cpp and turtle.hpp, you can see that I use a very simple model of a nonholonomic vehicle with forward velocity input and steering angle velocity input. 
Based on this you can extend the dynamic equations to more sophisticated models,
so that you can also control the drift of the car :-).
If you want to have some material about autonomous driving with NMPC, 
check out Mr. Alamdari's work:
https://wwwen.uni.lu/snt/people/seyed_amin_sajadi_alamdari 
To sum up:
1. Try to formulate the cars behaviour in terms of ODE's => Systemfunction f(x,u) and derive the the corresponding equations needed for the MPC (see turtle.cpp)
   Df(x,u)/Dx, Df(x,u)/Du, lambda'*Df(x,u)/Dx, etc.
2. Update the trajectory tracking function => l(x,u) and compute its derivatives
   dl(x,u)/dx, dl(x,u)/du (see turtle.cpp)
3. Update the interface, how you read and send data to your car (see the update routines in turtle.h)
4. Compile the package and run it on the real car.

If you have more specific questions, let me know!
Best regards, 
Jan

**Comment by caotangyu** [comment]
Thank you, Jan.
The code of tutorial_turtlesim branch is same with the master branch. There is one thing I want to confine that the  inequality constraints need to be transferred to sigmoid function ? Right ?  The sigmoid function can be part of cost function L  ?

**Comment by DentOpt** [comment]
Hi will,
the basic framework is the same,
but the underlying model of the turtlebot is a very simple representation of car dynamics.
For constraint handling there are several options:
1. The Ohtsuka auxiliary variable inequality constraint handling package 
(Option 1 of the Controller Class )
2. Logarithmic barrier constraint handling (Option 0 of the Controller Class )
3. The method you mention to implement your own constraint handling using sigmoid, tanh, in the cost function l etc.
It depends on your application.

For starters I would go with 3, as 1. and 2. need some parametrization to not lead to an infeasible solution. 3. has the disadvantage that it is a weakened constraint that allows the violation of the constraint. However, for most applications in robotics this is acceptable when using enough safety distance...

**Comment by caotangyu** [comment]
Hi, Jan
The tutorial_turtlesim branch  doesn’t have turtle.cpp and turtle.hpp  file. Could you tell me what each element of the array x[9] of system function f(x,u) described in the Ardrone_20170227.cpp means? I read the paper “A real-time model predictive position control with collision avoidance for commercial low-cost quadrotors”. I guess the System function f(x,u) should be:
[cid:image001.png@01D39E85.87829F80]
In this function X only have 6  element, but not 9. This confused me a lot.


祝您身体健康，工作顺利

发件人: DentOpt<mailto:notifications@github.com>
发送时间: 2017年12月14日 18:39
收件人: DentOpt/denmpc<mailto:denmpc@noreply.github.com>
抄送: will1991<mailto:tomwill1991@outlook.com>; Author<mailto:author@noreply.github.com>
主题: Re: [DentOpt/denmpc] compile error (#1)


Hi will,
the basic framework is the same,
but the underlying model of the turtlebot is a very simple representation of car dynamics.
For constraint handling there are several options:

  1.  The Ohtsuka auxiliary variable inequality constraint handling package
(Option 1 of the Controller Class )
  2.  Logarithmic barrier constraint handling (Option 0 of the Controller Class )
  3.  The method you mention to implement your own constraint handling using sigmoid, tanh, in the cost function l etc.
It depends on your application.

For starters I would go with 3, as 1. and 2. need some parametrization to not lead to an infeasible solution. 3. has the disadvantage that it is a weakened constraint that allows the violation of the constraint. However, for most applications in robotics this is acceptable when using enough safety distance...

―

**Comment by DentOpt** [comment]
Hi Will,

thanks a lot for your feedback.
There was a mixup in the branches. I updated the turtlesim branch. It now contains the Turtlesim files.
So I still recommend to go with that, as the model is more intuitive.

Regarding the Ardrone Model,
you are completely right, that the here included model has already 9 states.
This is already an updated model version, which is currently under publication
The states are:
position: x,y,z,    
orientation: dx,dy, (direction vector)
linear velocity: vx,vy,vz,
angular velocity: vyaw.

Thanks again and best regards!

**Comment by LisaWeichsel** [comment]
Using catkin build I still have the same output error :  Scheduler.h:17: error: ros/ros.h: No such file or directory

**Comment by benjaminabruzzo** [comment]
Do you have any launch files that you've used to actually control your AR.Drone?  I'm interested in first repeating your experiment for my own research.  I am also using an AR.Drone in an Optitrack system, so I was hoping to use as much of your previous configurations as possible.

**Comment by DentOpt** [comment]
Hi,
I have been using the Ardrone autonomy package.
http://wiki.ros.org/ardrone_autonomy
To make it work with denmpc, you just have to make sure to match the topic names of both packages.
For the optitrack system, I have used the mocap package:
https://github.com/ros-drivers/mocap_optitrack
Please don't hesitate to contact me, if you have any problems.
Best regards,
Jan

**Comment by LisaWeichsel** [comment]
Hi,
I would like to use this package to first simulate the closed-loop behavior of a car. I'm using a urdf model and gazebo-plugin to simulate the cars behavior.
Afterwards I want to control the real vehicle with the model predictive controller designed in the simulation. 
Has anyone experienced using the package in this context ?
Best regards,
Lisa

**Comment by DentOpt** [comment]
Hi,
[Seyed Amin Sajadi Alamdari](https://wwwen.uni.lu/snt/people/seyed_amin_sajadi_alamdari) is using MPC (CMSCGMRES) for adaptive cruise control of cars. His publications do also include a dynamical model of the lateral movement of a car.

As DENMPC can directly communicate with gazebo via ROS, this is easy to realise. 
The only thing you need to implement in DENMPC, is a dynamic model of your car and the objective function you want to use. If a simple kinematic model is sufficient for your purposes and you only want to control the position, you can just modify the turtlebot model [(turtlebot)](https://github.com/DentOpt/denmpc/blob/tutorial_turtlesim/src/user_generated/Turtle.cpp) to fit your car properties and that's it.

Let me know, if you need any help.
Best regards,
Jan

**Comment by LisaWeichsel** [comment]
Thanks.
Is it possible that the link is missing?
Best regards,
Lisa
